server/conn.go

```diff
@@ -1295,6 +1317,80 @@ func (cc *clientConn) handleQuery(ctx context.Context, sql string) (err error) {
 	return err
 }
 
+// prefetchPointPlanKeys extracts the point keys in multi-statement query,
+// use BatchGet to get the keys, so the values will be cached in the snapshot cache, save RPC call cost.
+func (cc *clientConn) prefetchPointPlanKeys(stmts []ast.StmtNode) ([]plannercore.Plan, error) {
+	txn, err := cc.ctx.Txn(false)
+	if err != nil {
+		return nil, err
+	}
+	if !txn.Valid() {
+		// Only build in-transaction query for simplicity.
+		// Later we can support out-transaction multi-statement query.
+		return nil, nil
+	}
+	vars := cc.ctx.GetSessionVars()
+	if vars.TxnCtx.IsPessimistic && vars.TxnCtx.Isolation == ast.ReadCommitted {
+		// ReadCommitted isolation update ForUpdateTS for every statement, the snapshot cache would be invalidated.
+		// Later we can support it by avoid update ForUpdateTS for multi-statement query.
+		return nil, nil
+	}
+	pointPlans := make([]plannercore.Plan, len(stmts))
+	var idxKeys []kv.Key
+	var rowKeys []kv.Key
+	is := domain.GetDomain(cc.ctx).InfoSchema()
+	sc := vars.StmtCtx
+	for i, stmt := range stmts {
+		if err = plannercore.Preprocess(cc.ctx, stmt, is); err != nil {
+			return nil, err
+		}
+		p := plannercore.TryFastPlan(cc.ctx.Session, stmt)
+		pointPlans[i] = p
+		if p == nil {
+			continue
+		}
+		// Only support Update for now.
+		// TODO: support other point plans.
```

Comment:
I think we could support select/delete. It won't be very complex.

Reply:
I'm trying to make the change as minimum as possible to make the review work easier.